### PR TITLE
 Add auto-grad support for index_select op

### PR DESCRIPTION
**Pull request by chang-l**
This PR adds the auto-grad support for `index_select`, to fuse it in fwd pass. I will work on bwd pass to enable `index_select` fusion within `torch.nn.module` implementation.

### CODE REVIEW

File: test/test_jit_cuda_fuser.py [hunk 1]
```diff
@@ -4185,6 +4185,47 @@ def t(x_kj, idx_kj, sbf):
...
+        jit_o = t_jit(x, y, ind)
+        jit_o.backward(grad)
+        x.grad.zero_()
+        y.grad.zero_()
+        jit_o = t_jit(x, y, ind)
+        jit_o.backward(grad)
+        self.assertEqual(ref_x.grad, x.grad)
+        self.assertEqual(ref_y.grad, y.grad)
+        self.assertTrue(self._compare("comparing output failed", o, jit_o, 1e-5))
+        self.assertGraphContainsExactly(t_jit.graph_for(x, y, ind), FUSION_GUARD, 1, consider_subgraphs=True)
```
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py]: nitpick: can we switch this from using `_run_trainign_helper` instead?

in that function, we are not checking grad, let's add that as well. :bow:

File: torch/csrc/jit/runtime/symbolic_script.cpp [hunk 2]
```diff
@@ -222,6 +222,17 @@ const std::vector<std::string> functions = {
...
             return torch.zeros(sizes).to(grad).scatter_(dim, indices, grad)
 
+        def index_select(self,
+                         dim: int,
+                         index):
+            output = torch.index_select(self, dim, index)
+            self_size = self.size()
+
+            def backward(grad_output):
+                grad_self = torch.zeros_like(self, memory_format=1).index_add_(dim, index, grad_output)
```
Comment (jjsjann123) [inline torch/csrc/jit/runtime/symbolic_script.cpp]: nitpick: `index_add_` -> `index_add`. Let's not encouraging in-place variants.

**Review by jjsjann123 — COMMENTED** [review]
LGTM, let's also push this to upstream directly :)

File: test/test_jit_cuda_fuser.py (hunk 1 above)
Comment (chang-l) [reply test/test_jit_cuda_fuser.py]: I updated `_run_training_helper` function to include grad checking. As it is called by `test_gelu`, and `test_dropout_training`, I have tested them all to make sure I didn't break anything.

**Review by jjsjann123 — APPROVED** [review]
LGTM~~ thx for cleaning up the tests~~

**Pull request merged**